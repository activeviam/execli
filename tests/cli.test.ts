import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";

import { describe, expect, test } from "@jest/globals";
import { execa } from "execa";
import { packageDirectorySync } from "pkg-dir";
import { temporaryFileTask } from "tempy";

const { bin } = JSON.parse(readFileSync("package.json", "utf8")) as Readonly<{
  bin: string;
}>;
const packageRootDirectory = packageDirectorySync()!;
const binPath = join(packageRootDirectory, bin);
const commandsFilePath = join(
  dirname(binPath),
  "__test_resources__",
  "commands",
  "index.js",
);

test("compile", async () => {
  // Use a temporary file path outside the package to make sure all the dependencies
  // are correctly bundled and are not required through the package's node_modules.
  await temporaryFileTask(
    async (targetPath) => {
      await execa("node", [binPath, "compile", commandsFilePath, targetPath]);
      const { stdout } = await execa(targetPath, [
        "simple",
        "--concurrency",
        "0",
      ]);
      expect(stdout).toContain("Echo middle");
      expect(stdout).toContain("mystery");
    },
    { extension: "mjs" },
  );
}, 15_000);

describe("run", () => {
  const runArguments = [binPath, "run", commandsFilePath];

  test.each([
    [
      "command required",
      {
        failing: true,
      } as const,
    ],
    [
      "option (existing)",
      {
        commandName: "simple",
        options: ["--flag"],
      },
    ],
    [
      "option (unexisting)",
      {
        commandName: "simple",
        failing: true,
        options: ["--unexisting-option"],
      },
    ],
    [
      "dryRun",
      {
        commandName: "simple",
        options: ["--dryRun"],
      },
    ],
    [
      "concurrent",
      {
        commandName: "simple",
        concurrency: 1,
      },
    ],
    [
      "from",
      {
        commandName: "simple",
        options: ["--from", "Echo middle"],
      },
    ],
    [
      "only (exact title)",
      {
        commandName: "simple",
        options: ["--only", "Echo start"],
      },
    ],
    [
      "only (slug)",
      {
        commandName: "simple",
        options: ["--only", "echo-start"],
      },
    ],
    [
      "only (neither exact title nor slug)",
      {
        commandName: "simple",
        failing: true,
        options: ["--only", "EcHo-StArt"],
      },
    ],
    [
      "skip (exact title)",
      {
        commandName: "simple",
        options: ["--skip", "Echo start"],
      },
    ],
    [
      "skip (slug)",
      {
        commandName: "simple",
        options: ["--skip", "echo-start"],
      },
    ],
    [
      "skip (neither exact title nor slug)",
      {
        commandName: "simple",
        failing: true,
        options: ["--skip", "EcHo-StArt"],
      },
    ],
    [
      "tag (existing)",
      {
        commandName: "simple",
        options: ["--tag", "a"],
      },
    ],
    [
      "tag (unexisting)",
      {
        commandName: "simple",
        failing: true,
        options: ["--tag", "d"],
      },
    ],
    [
      "until",
      {
        commandName: "simple",
        options: ["--until", "Echo middle"],
      },
    ],
    [
      "failed regular task",
      {
        commandName: "error",
        failing: true,
      },
    ],
    [
      "failed command task",
      {
        commandName: "error",
        failing: true,
        options: ["--from", "Run failing script"],
      },
    ],
    [
      "rollback",
      {
        commandName: "rollback",
        failing: true,
      },
    ],
  ])(
    "%s %j",
    async (
      _,
      {
        commandName,
        concurrency = 0,
        failing = false,
        options = [],
      }: Readonly<{
        commandName?: string;
        concurrency?: number;
        failing?: boolean;
        options?: readonly string[];
      }> = {},
    ) => {
      const result = await execa(
        "node",
        [
          ...runArguments,
          ...(commandName ? ["--", commandName] : []),
          "--concurrency",
          String(concurrency),
          ...options,
        ],
        {
          reject: !failing,
        },
      );

      expect(result.failed).toBe(failing);

      const outputs: Record<string, readonly string[]> = {};

      for (const outputType of ["stderr", "stdout"] as const) {
        if (result[outputType]) {
          outputs[outputType] = result[outputType]
            .replace(
              /(file:\/\/)?[\\\-/:.\w]*([/\\])execli[\\\-/:.\w]*/g,
              "SOME_EXECLI_PATH",
            )
            .replace(/node:[\\/\-:.\w]*/g, "SOME_NODE_PATH")
            .split(/\r?\n/);
        }
      }

      expect(outputs).toMatchSnapshot();
    },
    15_000,
  );
});
