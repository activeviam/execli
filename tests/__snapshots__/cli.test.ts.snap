// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run command required {"failing":true} 1`] = `
{
  "stderr": [
    "bin.js run <path>",
    "",
    "Run the commands at the given path, forwarding the command line arguments after \`--\`.",
    "",
    "Positionals:",
    "  path  The path resolving to the file exporting the commands.  [string] [required]",
    "",
    "Options:",
    "  --help  Show help  [boolean]",
    "",
    "Unknown argument: concurrency",
  ],
}
`;

exports[`run concurrent {"commandName":"simple","concurrency":1} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] Echo start: $ echo start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] Echo middle: $ echo mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] Echo end: $ echo end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run dryRun {"commandName":"simple","options":["--dryRun"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[TITLE] Echo start ($ echo start)",
    "[SKIPPED] Skipped by --dryRun option",
    "[STARTED] Echo middle",
    "[TITLE] Echo middle ($ echo mystery)",
    "[SKIPPED] Skipped by --dryRun option",
    "[STARTED] Echo end",
    "[TITLE] Echo end ($ echo end)",
    "[SKIPPED] Skipped by --dryRun option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run failed command task {"commandName":"error","failing":true,"options":["--from","Run failing script"]} 1`] = `
{
  "stderr": [
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "Error: Command failed with exit code 1: node --eval console.log('failing'); process.exit(1)",
    "    at ExecError.toDetailedError (SOME_EXECLI_PATH)",
    "    at runTask (SOME_EXECLI_PATH)",
    "    at process.processTicksAndRejections (SOME_NODE_PATH)",
    "    at async Object.handler (SOME_EXECLI_PATH)",
  ],
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Print current directory",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Nest task",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Run failing script",
    "[DATA] $ node --eval "console.log('failing'); process.exit(1)"",
    "failing",
  ],
}
`;

exports[`run failed regular task {"commandName":"error","failing":true} 1`] = `
{
  "stderr": [
    "[FAILED] Something went wrong.",
    "[FAILED] Something went wrong.",
    "[FAILED] Something went wrong.",
    "Error: Something went wrong.",
    "    at run (SOME_EXECLI_PATH)",
    "    at runRegularTask (SOME_EXECLI_PATH)",
    "    at Task.task (SOME_EXECLI_PATH)",
    "    at Task.run (SOME_EXECLI_PATH)",
    "    at async SOME_EXECLI_PATH",
  ],
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Print current directory",
    "[DATA] $ pwd",
    "SOME_EXECLI_PATH",
    "[SUCCESS] Print current directory",
    "[STARTED] Nest task",
    "[STARTED] Slow successful task",
    "[DATA] Slow but successful",
    "[SUCCESS] Slow successful task",
    "[STARTED] Throw exception",
  ],
}
`;

exports[`run from {"commandName":"simple","options":["--from","Echo middle"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Echo middle",
    "[DATA] $ echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] $ echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run only (exact title) {"commandName":"simple","options":["--only","Echo start"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] $ echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --only option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --only option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run only (neither exact title nor slug) {"commandName":"simple","failing":true,"options":["--only","EcHo-StArt"]} 1`] = `
{
  "stderr": [
    "Invalid values:",
    "  Argument: only, Given: "EcHo-StArt", Choices: "Echo end", "Echo middle", "Echo start", "Parent task"",
  ],
}
`;

exports[`run only (slug) {"commandName":"simple","options":["--only","echo-start"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] $ echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --only option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --only option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run option (existing) {"commandName":"simple","options":["--flag"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] $ echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] $ echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] $ echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run option (unexisting) {"commandName":"simple","failing":true,"options":["--unexisting-option"]} 1`] = `
{
  "stderr": [
    "Unknown arguments: unexisting-option, unexistingOption",
  ],
}
`;

exports[`run rollback {"commandName":"rollback","failing":true} 1`] = `
{
  "stderr": [
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "[ROLLBACK] Parent task",
    "Error: Parent task",
    "    at Task.run (SOME_EXECLI_PATH)",
    "    at process.processTicksAndRejections (SOME_NODE_PATH)",
    "    at async SOME_EXECLI_PATH",
  ],
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Run failing script",
    "[DATA] $ node --eval "console.log('failing'); process.exit(1)"",
    "failing",
    "[DATA] $ echo rollback",
    "rollback",
  ],
}
`;

exports[`run skip (exact title) {"commandName":"simple","options":["--skip","Echo start"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --skip option",
    "[STARTED] Echo middle",
    "[DATA] $ echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] $ echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run skip (neither exact title nor slug) {"commandName":"simple","failing":true,"options":["--skip","EcHo-StArt"]} 1`] = `
{
  "stderr": [
    "Invalid values:",
    "  Argument: skip, Given: "EcHo-StArt", Choices: "Echo end", "Echo middle", "Echo start", "Parent task"",
  ],
}
`;

exports[`run skip (slug) {"commandName":"simple","options":["--skip","echo-start"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --skip option",
    "[STARTED] Echo middle",
    "[DATA] $ echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] $ echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run tag (existing) {"commandName":"simple","options":["--tag","a"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] $ echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --tag option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --tag option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run tag (unexisting) {"commandName":"simple","failing":true,"options":["--tag","d"]} 1`] = `
{
  "stderr": [
    "Invalid values:",
    "  Argument: tag, Given: "d", Choices: "a", "b", "c"",
  ],
}
`;

exports[`run until {"commandName":"simple","options":["--until","Echo middle"]} 1`] = `
{
  "stdout": [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] $ echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] $ echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --until option",
    "[SUCCESS] Parent task",
  ],
}
`;
